src/InjectCoordinatesIntrinsics.cpp

```diff
@@ -0,0 +1,156 @@
+#include "InjectCoordinatesIntrinsics.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "CodeGen_GPU_Dev.h"
+#include "Substitute.h"
+#include "FuseGPUThreadLoops.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::string;
+using std::vector;
+
+class InjectCoordinatesIntrinsics : public IRMutator {
+public:
+    InjectCoordinatesIntrinsics() : inside_kernel_loop(false) {}
+    Scope<int> scope;
+    bool inside_kernel_loop;
+
+private:
+    using IRMutator::visit;
+
+    void visit(const Provide *provide) {
+        if (!inside_kernel_loop) {
+            IRMutator::visit(provide);
+            return;
+        }
+
+        internal_assert(provide->values.size() == 1)
+            << "Coordinate currently only supports single-valued stores.\n";
+        user_assert(provide->args.size() == 3)
+            << "Coordinate stores require three coordinates.\n";
+
+        // Create coordinate_store("name", name.buffer, x, y, c, value)
+        // intrinsic.
+        Expr value_arg = mutate(provide->values[0]);
+        vector<Expr> args = {
+            provide->name,
+            Variable::make(Handle(), provide->name + ".buffer"),
+            provide->args[0],
+            provide->args[1],
+            provide->args[2],
+            value_arg};
+
+        stmt = Evaluate::make(Call::make(value_arg.type(),
+                                         Call::coordinates_store,
+                                         args,
+                                         Call::Intrinsic));
+    }
+
+    void visit(const Call *call) {
+        if (!inside_kernel_loop || call->call_type == Call::Intrinsic ||
+            call->call_type == Call::Extern) {
+            IRMutator::visit(call);
+            return;
+        }
+
+        string name = call->name;
+        if (call->call_type == Call::Halide && call->func.outputs() > 1) {
+            name = name + '.' + int_to_string(call->value_index);
+        }
+
+        vector<Expr> call_args = call->args;
+        // Check to see if we are reading from a one or two dimension function
+        // and pad to three dimensions.
+        while (call_args.size() < 3) {
+            call_args.push_back(IntImm::make(0));
+        }
+
+        // Create coordinates_load("name", "name[.n]", name.buffer, x, y, c)
```

Comment:
Passing value_index won't be sufficient because it's actually "name[.n]":
```
        if (call->call_type == Call::Halide && call->func.outputs() > 1) {
            name = name + '.' + int_to_string(call->value_index);
        }
```